pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+        offsets = [0] + list(offsets)
+        loop_indices = [range(offsets[i], o) for i, o in enumerate(offsets[1:])]
+        self._info = [{'loop_indices': li} for li in loop_indices]
+
+    def __call__(self, loop_chain):
+        fused_par_loops = []
+        for kernel, info in zip(self._kernel, self._info):
+            loop_indices = info['loop_indices']
+            extra_args = info.get('extra_args')
+            # Create the ParLoop's arguments. Note that both the iteration set and
+            # the iteration region must correspond to that of the /base/ loop
+            iterregion = loop_chain[loop_indices[0]].iteration_region
+            iterset = loop_chain[loop_indices[0]].it_space.iterset
+            loops = [loop_chain[i] for i in loop_indices]
+            args = Arg.filter_args([loop.args for loop in loops]).values()
+            # Create any ParLoop's additional arguments
+            extra_args = [Dat(*d)(*a) for d, a in extra_args] if extra_args else []
+            args += extra_args
+            # Create the actual ParLoop, resulting from the fusion of some kernels
+            fused_par_loops.append(_make_object('ParLoop', kernel, iterset, *args,
+                                                **{'iterate': iterregion}))
+        return fused_par_loops
+
+
+class HardFusionSchedule(FusionSchedule):
+    """Schedule a sequence of :class:`ParLoop` objects after hard fusion."""
+
+    def __init__(self, schedule, fused):
+        self._schedule = schedule
+        self._fused = fused
+
+        # Set proper loop_indices for this schedule
+        self._info = dcopy(schedule._info)
+        for i, info in enumerate(schedule._info):
+            for k, v in info.items():
+                self._info[i][k] = [i] if k == 'loop_indices' else v
+
+        # Update the input schedule to make use of hard fusion kernels
+        kernel = scopy(schedule._kernel)
+        for fused_kernel, fused_map in fused:
+            base, fuse = fused_kernel._kernels
+            base_idx, fuse_idx = kernel.index(base), kernel.index(fuse)
+            pos = min(base_idx, fuse_idx)
+            kernel.insert(pos, fused_kernel)
+            self._info[pos]['loop_indices'] = [base_idx, fuse_idx]
+            # In addition to the union of the /base/ and /fuse/'s sets of arguments,
+            # a bitmap, with i-th bit indicating whether the i-th iteration in "fuse"
+            # has been executed, will have to be passed in
+            self._info[pos]['extra_args'] = [((fused_map.toset, None, np.int32),
+                                              (RW, fused_map))]
+            kernel.pop(pos+1)
+            pos = max(base_idx, fuse_idx)
+            self._info.pop(pos)
+            kernel.pop(pos)
+        self._kernel = kernel
+
+    def __call__(self, loop_chain, only_hard=False):
+        # First apply soft fusion, then hard fusion
+        if not only_hard:
+            loop_chain = self._schedule(loop_chain)
```

Comment:
The schedules are applied incrementally. Given a loop chain, you first apply soft fusion, then hard fusion (then tiling is specifically requested by the user, but this is another story). 

``only_hard`` is to invoke the scheduler without performing soft_fusion on the provided loop chain because you already know that it's been performed already. This is the case when you have just created a hard fusion ParLoop, and you want to update the (already soft fused) loop chain